arm-runtimes/CMakeLists.txt

```diff
@@ -0,0 +1,795 @@
+set(ENABLE_LIBC_TESTS ${ENABLE_LIBC_TESTS_def} CACHE BOOL "Enable libc tests (picolibc, newlib or llvm-libc).")
+set(ENABLE_COMPILER_RT_TESTS ${ENABLE_COMPILER_RT_TESTS_def} CACHE BOOL "Enable compiler-rt tests.")
+set(ENABLE_LIBCXX_TESTS ${ENABLE_LIBCXX_TESTS_def} CACHE BOOL "Enable libcxx tests.")
+set(LLVM_BINARY_DIR "" CACHE PATH "Path to LLVM toolchain root to build libraries with")
+set(LIBC_HDRGEN "" CACHE PATH "Path to prebuilt lbc-hdrgen if not included in LLVM binaries set by LLVM_BINARY_DIR")
+
+# Temporary location to collect the libraries as they are built.
+set(TEMP_LIB_DIR "${CMAKE_CURRENT_BINARY_DIR}/tmp_install")
+
+include(ExternalProject)
+include(${TOOLCHAIN_SOURCE_DIR}/cmake/fetch_llvm.cmake)
+
+# For building picolibc use Meson.
+# Although picolibc has support for building with CMake, the Meson code
+# is more mature and works better with LLVM.
+find_program(MESON_EXECUTABLE meson REQUIRED)
+find_package(Python3 REQUIRED COMPONENTS Interpreter)
+
+# If a compiler launcher such as ccache has been set, it should be
+# passed down to each subproject build.
+set(compiler_launcher_cmake_args "")
+if(CMAKE_C_COMPILER_LAUNCHER)
+    list(APPEND compiler_launcher_cmake_args "-DCMAKE_C_COMPILER_LAUNCHER=${CMAKE_C_COMPILER_LAUNCHER}")
+endif()
+if(CMAKE_CXX_COMPILER_LAUNCHER)
+    list(APPEND compiler_launcher_cmake_args "-DCMAKE_CXX_COMPILER_LAUNCHER=${CMAKE_CXX_COMPILER_LAUNCHER}")
+endif()
+
+if(TARGET_ARCH MATCHES "^aarch64")
+    set(target_triple "aarch64-none-elf")
+    set(cpu_family aarch64)
+else()
+    # Choose the target triple so that compiler-rt will do the
+    # right thing. We can't always put the exact target
+    # architecture in the triple, because compiler-rt's cmake
+    # system doesn't recognize every possible Arm architecture
+    # version. So mostly we just say 'arm' and control the arch
+    # version via -march=armv7m (or whatever).
+    # Exceptions are architectures pre-armv7, which compiler-rt expects to
+    # see in the triple because that's where it looks to decide whether to
+    # use specific assembly sources.
+    if(TARGET_ARCH MATCHES "^armv[4-6]")
+        set(target_triple "${TARGET_ARCH}-none-eabi")
+    else()
+        set(target_triple "arm-none-eabi")
+    endif()
+    if(COMPILE_FLAGS MATCHES "-mfloat-abi=hard")
+        # Also, compiler-rt looks in the ABI component of the
+        # triple to decide whether to use the hard float ABI.
+        set(target_triple "${target_triple}hf")
+    endif()
+    set(cpu_family arm)
+endif()
+
+# Create a single target for all testing. If no testing is enabled, this
+# will simply do nothing.
+add_custom_target(check-all)
+
+# If any testing is enabled, prepare test executor settings.
+if(ENABLE_LIBC_TESTS OR ENABLE_COMPILER_RT_TESTS OR ENABLE_LIBCXX_TESTS)
+    # Flags required to link tests.
+    set(test_link_flags "-nostartfiles -lcrt0-semihost -lsemihost -T picolibcpp.ld")
```

Comment:
I'm still confused about these flags being picolibc-specific. One of the conditions that enables this if statement is `ENABLE_LIBC_TESTS`. But that variable is documented further up this file as `"Enable libc tests (picolibc, newlib or llvm-libc)."`

Either we can run tests on other libcs or we can't, surely? If we can't then there should be a check here that throws an error; if we can then this should be an if statement that sets up appropriate flags?

Reply:
There are checks to throw an error for attempting to enable testing with anything other than picolibc selected, but it's done later in the library specific blocks.

But you are right in that it should be clear that these will also need updating when we start adding more testing. Additionally throwing an error here is probably the safest way to ensure the currently hardcoded flags don't get missed.